### ISSUE TITLE
SA1316 should not be reported for tuple deconstruction

### ISSUE THREAD

**Issue by Arthri** [DotNetAnalyzers/StyleCopAnalyzers#3878]
Tuples can follow two different casing conventions depending on how it's used: tuple types may follow pascal casing, while deconstruction may follow camel casing

The rule does not differentiate between the two uses and indiscriminately flags both, so I propose having a separate rule for tuple deconstruction

```cs
// no diagnostic
(string K, string V) a = ("", "");
// SA1316
(string k, string v) = a;

// SA1316
foreach ((string k, string v) in dict)
{
}
```

**Comment by bjornhellander** [comment]
I am not seeing this behavior. Which version are you using?

**Comment by bjornhellander** [comment]
Have you set includeInferredTupleElementNames to true in your stylecop.json configuration file? I see the behavior that you describe when I do that.

**Comment by bjornhellander** [comment]
`var tuple = (x, y)` is equivalent to `var tuple = (x: x, y: y),`. Setting `includeInferredTupleElementNames` to true means that SA1316 will also check the implicit/inferred element names if they are omitted, like they are in `var tuple = (x, y)`. So with the default PascalCase configuration, that code would also trigger a SA1316 warning.

**Comment by bjornhellander** [comment]
It looks to me like you have found a bug. In my opinion, tuple expressions used in the "target" of a deconstruction should not be checked by this rule. @sharwell, do you agree? I am a bit surprised that this hasn't been found before, but `includeInferredTupleElementNames` is unfortunately false by default, so maybe it isn't used much.

**Comment by bjornhellander** [comment]
It looks like setting `includeInferredTupleElementNames` to false doesn't work well either:
This rule does not even trigger on `var tuple = (x: 1, Y: 2)`, which I assume it should.

**Comment by MattFromRVA** [comment]
@bjornhellander FYI I believe you fixed the rule not triggering on `var tuple = (x: 1, Y: 2)` with bjornhellander/NewStyleCopAnalyzers#3.

Rule now triggers with latest version of [NewStyleCopAnalyzers](https://github.com/bjornhellander/NewStyleCopAnalyzers).

I am creating a PR to fix the other issues.
